src/contracts/tokens/RewardPool.sol

```diff
@@ -0,0 +1,237 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity 0.8.20;
+
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+import {Authorizable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {Modifiable} from "@contracts/utils/Modifiable.sol";
+
+import {IRewardPool} from "@interfaces/tokens/IRewardPool.sol";
+import {Math, RAY, WAD} from "@libraries/Math.sol";
+
+/**
+ * @title  RewardPool
+ * @notice This contract constitutes a reward pool for a given reward token
+ */
+contract RewardPool is Authorizable, Modifiable, IRewardPool {
+    using Math for uint256;
+    using SafeERC20 for IERC20;
+
+    // --- Registry ---
+
+    /// @inheritdoc IRewardPool
+    IERC20 public immutable rewardToken;
+
+    // --- Params ---
+
+    /// @inheritdoc IRewardPool
+    // solhint-disable-next-line private-vars-leading-underscore
+    RewardPoolParams public _params;
+
+    /// @inheritdoc IRewardPool
+    function params() external view returns (RewardPoolParams memory _params) {
+        return _params;
+    }
+
+    // --- Data ---
+
+    /// @inheritdoc IRewardPool
+    uint256 private _totalStaked;
+    /// @inheritdoc IRewardPool
+    uint256 public rewardPerTokenStored;
+    /// @inheritdoc IRewardPool
+    uint256 public periodFinish = 0;
+    /// @inheritdoc IRewardPool
+    uint256 public rewardRate = 0;
+    /// @inheritdoc IRewardPool
+    uint256 public lastUpdateTime;
+    /// @inheritdoc IRewardPool
+    uint256 public queuedRewards = 0;
+    /// @inheritdoc IRewardPool
+    uint256 public currentRewards = 0;
+    /// @inheritdoc IRewardPool
+    uint256 public historicalRewards = 0;
+    /// @inheritdoc IRewardPool
+    uint256 public rewardPerTokenPaid = 0;
+    /// @inheritdoc IRewardPool
+    uint256 public rewards = 0;
+
+    // --- Init ---
+
+    /**
+     * @param _rewardToken Address of the reward token
+     */
+    constructor(address _rewardToken) {
+        if (_rewardToken == address(0)) revert RewardPool_InvalidRewardToken();
+        rewardToken = IERC20(_rewardToken);
+        _params.duration = 7 days;
+        _params.newRewardRatio = 830;
+    }
+
+    // --- Methods ---
+
+    /// @inheritdoc IRewardPool
+    function lastTimeRewardApplicable()
+        public
+        view
+        returns (uint256 _lastTime)
+    {
+        return Math.min(block.timestamp, periodFinish);
+    }
+
+    /// @inheritdoc IRewardPool
+    function totalStaked() public view returns (uint256) {
+        return _totalStaked;
+    }
+
+    /// @inheritdoc IRewardPool
+    function stake(
+        uint256 _wad
+    ) external updateReward(msg.sender) isAuthorized {
+        if (_wad == 0) revert RewardPool_StakeNullAmount();
+        _totalStaked += _wad;
+        emit RewardPool_Staked(msg.sender, _wad);
```

Comment:
this is being called by the StakingManager, so msg.sender will not be the end user, it will be the StakingManager.

Reply:
So the StakingManager only claims rewards from the reward pools. User's then claim these rewards from the StakingManager. So in this case msg.sender should be the StakingManager.